cheroot/connections.py

```diff
@@ -99,76 +99,87 @@ def expire(self):
         # find any connections still registered with the selector
         # that have not been active recently enough.
         threshold = time.time() - self.server.timeout
-        timed_out_connections = (
+        timed_out_connections = [
             (sock_fd, conn)
             for _, (_, sock_fd, _, conn) in self._selector.get_map().items()
             if conn != self.server and conn.last_used < threshold
-        )
+        ]
```

Comment:
i found that without capturing these to a separate list, i was seeing errors modifying the selector map during iteration in the `for sock_fd, conn in timed_out_connections` block below

Reply:
Yeah: https://github.com/cherrypy/cheroot/pull/309/files#r469927485 / #312.

But the proper fix would be to `.copy()` the dict early. This problem arises from the change in Python 3 (compared to Python 2): `.items()` returns a lazy iterator under py3 while in py2 it was a list.

Just turning this into a list may still hit this race condition because you're iterating over a moving target.

Using `.copy()` early should be atomic enough to prevent this and is recommended to mitigate such cases, in general.

Reply:
it's not clear to me that this is a threading issue - it's an error even in a single-threaded context to modify a dict during iteration, which is what's happening here.

in any event, `get_map()` returns a `collections.abc.Mapping` which does not support `copy()` - what would you prefer in this case? `copy.copy(self._selector.get_map()).items()`?

also - this code appears to be exercised by tests, i don't understand why it's not caught as an error there.

Reply:
https://github.com/cherrypy/cheroot/pull/313 if `copy.copy()` works for you

Reply:
Interesting... So probably in production envs selector map may change on the low-level more often, I guess.

Reply:
oh - no, i think it's just the simple fact that `self._selector.unregister()` modifies the selector map, and if that's called while iterating, an error is raised

Reply:
Yeah, looks like you're correct.